yt/frontends/amrvac/fields.py

```diff
@@ -29,14 +29,43 @@
     "spherical": ("r", "theta", "phi")
 }
 
+def _velocity(idir, idust=None):
+    """Generate a velocity function v_idir = m_idir / rho
+
+    If idust is not used, the "gas" velocity function is returned.
+
+    Parameters
+    ----------
+    idir : int
+        the direction index (1, 2 or 3)
+    idust : int, optional
+        the dust species index (AMRVAC). idust >= 1
+    """
+    if idust is None:
+        dust_label = ""
+    else:
+        dust_label = "dust%d_" % idust
+    def velocity_idir(field, data):
+        return data["gas", "%smoment_%d" % (dust_label, idir)] / data["gas", "%sdensity" % dust_label]
+    return velocity_idir
+
+code_density = "code_mass / code_length**3"
+code_moment = "code_mass / code_length**2 / code_time"
+code_pressure = "code_mass / code_length / code_time**2"
+
+# for now, define a finite family of dust fields (up to 100 species, should be enough)
+MAXN_DUST_SPECIES = 100
+known_dust_fields = [("rhod%d" % idust, (code_density, ["dust%d_density" % idust], None))
+                     for idust in range(1, MAXN_DUST_SPECIES+1)]
+for idir in range(1, 4):
+    known_dust_fields += [("m%dd%d" % (idir, idust), (code_moment, ["dust%d_moment_%d" % (idust, idir)], None))
+                          for idust in range(1, MAXN_DUST_SPECIES+1)]
+
```

Comment:
This will be always executed regardless if dust is present in the simulation or not, isn't it better to do this _only_ when there is dust?

Reply:
No, it's fine. `known_fields` is only a whitelist of fields we expect to find on disk with some info regarding the units it should be read in and how to display it and such...
`know_dust_fields` is just a intermediary variable I use to populate the `known_fields` list, which already contains fields that are not in every datfiles (m3, b1...)